nemo/utils/memory_profile_analyzer.py

```diff
@@ -0,0 +1,473 @@
+        data = pickle.load(f)
+    return data
+
+def make_hashable(obj):
+    """
+    Helper function to make an object hashable. 
+    """
+    if isinstance(obj, dict):
+        return tuple(sorted((k, make_hashable(v)) for k, v in obj.items()))
+    if isinstance(obj, list):
+        return tuple(make_hashable(e) for e in obj)
+    return obj
+
+def prune_frames(frames):
+    """
+    The stack trace is too long, and include many not-so-useful frame. We prune those frames with '??' in the filename. 
+    """
+    return [frame for frame in frames if '??' not in frame['filename']]
+
+def get_printed_frames(alive_memory):
+    """
+    Get the printed frames for the alive_memory. 
+    prune_frames to get rid of '??' in the filename. 
+    Add `\n` in between each frame for the readability. 
+    """
+    # Prune Frames
+    after_prune_frames = [prune_frames(x[3]) for x in alive_memory]
+    # add `\n` to each frame. Connect them together.
+    printed_frames = ['\n'.join([str(frame) for frame in frames]) for frames in after_prune_frames]
+    return printed_frames
+
+
+def read_tp(timepoint):
+    time_us = timepoint['time_us']
+    addr = timepoint['addr']
+    action = timepoint['action']
+    size = timepoint['size'] / GB_SIZE
+    frames = timepoint['frames']
+    stream = timepoint['stream']
+    return (time_us, addr, action, size, frames, stream)
+
+
+def first_py_frame(alloc_frames):
+    """
+    Return the frame when it first shows a `.py` file.
+    """
+    # target_string = 'layernorm_linear.py'
+    for frame in alloc_frames:
+        if '.py' in frame['filename']:
+            return frame
+    return None
+
+def alloc_memory_timeline(trace):
+    alloc_memory_history = np.zeros(len(trace)) # record the curr_alloc_memory at each timepoint
+    time_us_history = np.zeros(len(trace)) # record the time_us at each timepoint
+
+    curr_alloc_memory = 0 # in GB
+
+    min_alloc_memory = float('inf')
+    max_alloc_memory = float('-inf')
+
+    time_min_alloc = None
+    time_max_alloc = None
+    idx_min = None
+    idx_max = None
+    
+    for idx, timepoint in enumerate(trace):
+        (time_us, addr, action, size, frames, stream) = read_tp(timepoint)
+
+        if (action == "alloc"):
+            curr_alloc_memory += size
+        elif (action == "free_completed"):
+            curr_alloc_memory -= size
+
+        alloc_memory_history[idx] = curr_alloc_memory
+        time_us_history[idx] = time_us
+
+        if curr_alloc_memory > max_alloc_memory:
+            max_alloc_memory = curr_alloc_memory
+            time_max_alloc = time_us
+            idx_max = idx
+
+        if curr_alloc_memory < min_alloc_memory:
+            min_alloc_memory = curr_alloc_memory
+            time_min_alloc = time_us
+            idx_min = idx
+
+
+    return (alloc_memory_history, time_us_history), (idx_min, time_min_alloc, min_alloc_memory), (idx_max, time_max_alloc, max_alloc_memory)
+
+
+
+def record_alloc_memory_timeline(trace):
```

Comment:
this looks very similar to `alloc_memory_timeline` can you refactor to reduce duplicate code?